eventuals/lock.h

```diff
@@ -694,20 +694,41 @@ class ConditionVariable {
     return eventuals::Wait(
         lock_,
         [this, f = std::move(f), waiter = Waiter()](auto notify) mutable {
-          waiter.notify = std::move(notify);
-          if (head_ == nullptr) {
-            head_ = &waiter;
-          } else if (head_->next == nullptr) {
-            head_->next = &waiter;
+          // Check if template `F` is for `EmptyCondition`.
+          constexpr bool using_empty_condition = std::is_same_v<F, EmptyCondition>;
+
+          // Determine current state of condition variable.
+          // Note: Initializing using an `?:` did not work, so here goes verbatim.
+          bool current_condition;
+          if constexpr (using_empty_condition) {
+            current_condition = f(waiter);
           } else {
-            auto* next = head_->next;
-            while (next->next != nullptr) {
-              next = next->next;
+            current_condition = f();
+          }
+
+          if (current_condition == true) {
+            // If the current condition is `true`, the waiter need to be
+            // enqueued with other waiters so it can later be notified.
+
+            // Assign `nofify` callback to `waiter` for later use.
+            waiter.notify = std::move(notify);
+
+            // Add `waiter` to list of waiters. The below might look convoluted
+            // at first but is a text book "append" to a linked list.
+            if (head_ == nullptr) {
+              head_ = &waiter;
+            } else if (head_->next == nullptr) {
+              head_->next = &waiter;
+            } else {
+              auto* next = head_->next;
+              while (next->next != nullptr) {
+                next = next->next;
+              }
+              next->next = &waiter;
             }
-            next->next = &waiter;
           }
           return [&]() {
```

Comment:
And now based on comments from above this can become:
```suggestion
          return wait;
```